psdk_wrapper/src/modules/hms_utils.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2024 Unmanned Life
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/**
+ * @file hms_utils.hpp
+ *
+ * @brief Source file containing utility functions for the HMS module
+ *
+ * @authors Alejandro Mora
+ * Contact: alejandro@unmanned.life
+ *
+ */
+
+#include "psdk_wrapper/modules/hms_utils.hpp"
+
+namespace psdk_ros2
+{
+psdk_interfaces::msg::HmsInfoTable
+to_ros2_msg(const T_DjiHmsInfoTable& hms_info_table,
+            const nlohmann::json& codes, const char* language)
+{
+  psdk_interfaces::msg::HmsInfoTable ros2_hms;
+  ros2_hms.num_msg = hms_info_table.hmsInfoNum;
+  ros2_hms.table.resize(hms_info_table.hmsInfoNum);
+
+  for (int i = 0; i < hms_info_table.hmsInfoNum; i++)
+  {
+    // Extract error codes. If the "air" error code exists, it is
+    // implied that the associated "ground" error code also exists
+    std::string ground_key_lower_case =
+        "fpv_tip_" +
+        json_utils::to_hex_str(hms_info_table.hmsInfo[i].errorCode);
+    std::string ground_key_upper_case =
+        "fpv_tip_" +
+        json_utils::to_hex_str(hms_info_table.hmsInfo[i].errorCode, false);
+    std::string air_key_lower_case = ground_key_lower_case + "_in_the_sky";
+    std::string air_key_upper_case = ground_key_upper_case + "_in_the_sky";
+
+    // Hex error codes can be either lower or upper case, inspect both
+    auto ground_error_code = (codes.find(ground_key_lower_case) != codes.end()
+                                  ? codes.find(ground_key_lower_case)
+                                  : codes.find(ground_key_upper_case));
+    if (ground_error_code != codes.end())
+    {
+      auto ground_error_message = ground_error_code->find(language);
+      if (ground_error_message != ground_error_code->end())
+      {
+        ros2_hms.table[i].error_code = hms_info_table.hmsInfo[i].errorCode;
+        ros2_hms.table[i].component_index =
+            hms_info_table.hmsInfo[i].componentIndex + 1;
+        ros2_hms.table[i].error_level = hms_info_table.hmsInfo[i].errorLevel;
+        ros2_hms.table[i].ground_info = *ground_error_message;
+
+        auto air_error_code = (codes.find(air_key_lower_case) != codes.end()
+                                   ? codes.find(air_key_lower_case)
+                                   : codes.find(air_key_upper_case));
+        if (air_error_code != codes.end())
```

Comment:
Just a clarification, the air error code only exists if an associated ground error code exists and that is why this if is inside `if (ground_error_code != codes.end())`?